[RISCV] Add scheduler definitions for SpacemiT-X60

mikhailramalho opened the pull request:
This patch adds an initial scheduler model for the SpacemiT-X60, including latency for scalar instructions only.

The scheduler is based on the documented characteristics of the C908, which the SpacemiT-X60 is believed to be based on, and provides the expected latency for several instructions. I ran a probe to confirm all of these values and to get the latency of instructions not provided by the C908 documentation (e.g., double floating-point instructions).

For load and store instructions, the C908 documentation says the latency is \>= 3 for load and 1 for store. I tried a few combinations of values until I got the current values of 5 and 3, which yield the best results.

Although the X60 does appear to support multiple issue for at least some floating point instructions, this model assumes single issue as increasing it reduces the gains below.

This patch gives a geomean improvement of ~4% on SPEC CPU 2017 for both rva22u64 and rva22u64_v, with some benchmarks improving up to 18% (508.namd_r). There were a couple of execution time regressions, but only in noisy benchmarks (523.xalancbmk_r and 510.parest_r).

* rva22u64: https://lnt.lukelau.me/db_default/v4/nts/507?compare_to=405 (compares a55f7275 to the baseline 8286b804)
* rva22u64_v: https://lnt.lukelau.me/db_default/v4/nts/474?compare_to=404 (compares a55f7275 to the baseline 8286b804)

This initial scheduling model is strongly focused on providing sufficient definitions to provide improved performance for the SpacemiT-X60. Further incremental gains may be possible through a much more detailed microarchitectural analysis, but that is left to future work.

Further scheduling definitions for RVV can be added in a future PR.

mikhailramalho committed:
[RISCV] Add scheduler for x60

asb commented:
I've discussed earlier versions of this patch with Mikhail a fair bit so I'd appreciate a review from outside our org, but I will just share my thoughts on criteria for whether this makes sense to merge at this point:

For me the key thing is that per Mikhail's benchmarking it's at the point where there's pretty much across the board improvements and importantly, there's no evidence of regressions due to introducing scalar but not having vector scheduling. As for the fine details of the model, we're not going to get the very high fidelity of the 7-series model without much more microarchitectural information or a lot more reverse engineering. What is here seems a reasonable starting point.

I'd appreciate comments on anything that seems anomalous vs other scheduling models - we should aim to basically match the pattern of others unless there's data or documented information to differ. With that in mind, the store latency of 3 is a slight oddity vs other similar models. But the A55 had latency=4 up to f73334c46d59ffdbf12353932d5985049335a6c7 and the commit comment indicates it has very limited impact on scheduling anyway. So unless people feel strongly it's worth more investigation right now, I propose sticking with what Mikhail suggests.

(Incidentally, should any of our models be setting RetireOOO like the A55 does?)

mikhailramalho commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+def : WriteRes<WriteFMinMax64, [SMX60_FP]> { let Latency = 3; }
+
+let Latency = 10, ReleaseAtCycles = [10] in {
+  def :  WriteRes<WriteFDiv64, [SMX60_FP]>;
```
llvm-exegesis suggests the same latency for single- and double-precision fdiv. I just double-checked locally:
```
$ taskset -c 2 ./llvm-riscv/bin/llvm-exegesis --mode=latency --opcode-name=FDIV_D -mtriple=riscv64 -mcpu=spacemit-x60 --min-instructions=10000
  - { key: latency, value: 10.463, per_snippet_value: 10.463, validation_counters: {} }
$ taskset -c 2 ./llvm-riscv/bin/llvm-exegesis --mode=latency --opcode-name=FDIV_S -mtriple=riscv64 -mcpu=spacemit-x60 --min-instructions=10000
  - { key: latency, value: 10.4673, per_snippet_value: 10.4673, validation_counters: {} }
```
Across multiple runs, the latencies for both instructions range from 10.46xx to 10.49xx

topperc replied:
In my experience, llvm-exegesis can't accurately measure division latency on most hardware. The latency of division is commonly proportional to log2(dividend) - log2(divisor). llvm-exegesis doesn't control the values in the registers when it measures latency.

mikhailramalho replied:
I'll reconfirm with some handwritten fdiv latency-probing snippets and update here

preames reviewed (commented):
This looks general reasonable, and I agree that for a black box schedule model the right threshold is to use observed performance.  

I'm going to run through (in a second pass) the available data for this core, and cross check the model.  Forthcoming shortly.

topperc commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+
+// Integer division/remainder
+// Worst case latency is used.
+def : WriteRes<WriteIDiv32, [SMX60_IEU]> { let Latency = 12; }
```
I would be surprised if the CPU can issue 2 divides. I would also be surprised if the divider was pipelined. Probably needs `ResourceCycles` to be increased.

mshockwave replied:
I think you meant `ReleaseAtCycles`

topperc replied:
I did. I think ResourceCycles was the original name

zqb-all replied:
Yes, x60 can issue only one divide, and it's not pipelined

wangpc-pp commented:
I just wonder why we didn't ask Spacemit guys to provide the schedule model. They have a compiler team but are not so active in upstream.

mshockwave commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+def : WriteRes<WriteCTZ, [SMX60_IEU]>;
+def : WriteRes<WriteCTZ32, [SMX60_IEU]>;
+
+def : WriteRes<WriteCPOP, [SMX60_IEU]>;
```
could you double check if it actually has single-cycle CPOP?

mshockwave commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+}
+
+// Double precision
+def : WriteRes<WriteFAdd64, [SMX60_FP]> { let Latency = 4; }
+def : WriteRes<WriteFMul64, [SMX60_FP]> { let Latency = 4; }
```
nit: factor out `let Latency = 4` for these two entries? Same goes for FSGNJ64 and FMinMax64 below

mshockwave commented on llvm/test/tools/llvm-mca/RISCV/SpacemitX60/atomic.s:
```diff
@@ -0,0 +1,312 @@
+# NOTE: Assertions have been autogenerated by utils/update_mca_test_checks.py
+# RUN: llvm-mca -mtriple=riscv64 -mattr=+rva22u64 -mcpu=spacemit-x60 -iterations=1 < %s | FileCheck %s
```
I don't think you need `-mattr` here, the CPU definition should already specify all the extensions

mshockwave replied:
I would recommend and prefer all the new scheduling model tests to use the more concise `-instruction-tables=full` instead of the default MCA view

mshockwave commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+  def SMX60_LS : ProcResource<2>;
+
+  // An IEU can decode and issue two instructions at the same time
+  def SMX60_IEU : ProcResource<2>;
```
judging from other review comments related to this resource, should we split it into two ProcResource? And use ProcResGroup for entries that don't care about which pipe.

mikhailramalho committed:
Merge remote-tracking branch 'origin/main' into x60-sched

mikhailramalho committed:
Add comment about fp proc

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Split IEU procResource and set jump to only use one

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
single-issue for div and add ReleaseAtCycles

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Grouped together latencies with the same value

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Removed -mattr

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Added -instruction-tables=full to the tests

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Update latencies based on experiments

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

zqb-all commented:
> I just wonder why we didn't ask Spacemit guys to provide the schedule model. They have a compiler team but are not so active in upstream.
> 
> cc @zqb-all @zengdage @sunshaoce 

Thanks Mikhail for bringing the initial schedule model support to x60. We will take a look at this patch and work with the upstream to improve the performance of x60.

mikhailramalho committed:
Add latency for cpop

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

wangpc-pp commented on llvm/lib/Target/RISCV/RISCV.td:
```diff
@@ -57,6 +57,7 @@ include "RISCVSchedSyntacoreSCR345.td"
 include "RISCVSchedSyntacoreSCR7.td"
 include "RISCVSchedTTAscalonD8.td"
 include "RISCVSchedXiangShanNanHu.td"
+include "RISCVSchedSpacemitX60.td"
```
These files are in alphabetical order.

mikhailramalho committed:
Updated latency based on experiments

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Updated tests

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Reorder includes

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Fix div/rem latencies

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Updated test case

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
More test updates

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

zqb-all commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+  // An IEU can decode and issue two instructions at the same time
+  def SMX60_IEU : ProcResource<2>;
+
+  def SMX60_FP : ProcResource<1>;
```
Some floating-point instructions can double issue, such as those using FALU and FMAU, but not, for example, FCVT.
Mikhail mentioned that a value of 1 would give better performance, so we can start with 1. We will continue to improve this model in the future

preames replied:
Let's start with 1 here, and then see if we can split this in a follow up patch.

zqb-all commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+//===----------------------------------------------------------------------===//
+
+// Branching
+def : WriteRes<WriteJmp, [SMX60_IEU]>;
```
There is only one BJU, so only 1 branch instruction per cycle

zqb-all commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+def SpacemitX60Model : SchedMachineModel {
+  let IssueWidth        = 2; // dual-issue
+  let MicroOpBufferSize = 0; // in-order
+  let LoadLatency       = 5; // worse case: >= 3
```
Load latency is 3 or 4 in the case of cachehit, but since load=5 actually performs the best in tests, we can keep this until another configuration beats it in test performance

zqb-all commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,332 @@
+  let IssueWidth        = 2; // dual-issue
+  let MicroOpBufferSize = 0; // in-order
+  let LoadLatency       = 5; // worse case: >= 3
+  let MispredictPenalty = 9; // nine-stage
```
In the case of L1 cache hit, the penalty is about 3-6 cycle
However, we didn't test the performance impact of tuning this parameter. If a different value is better for the test results, then just use it :)

mikhailramalho committed:
Swap the order so comment makes more sense

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Revert back div/rem latencies

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Added tests for clmul, clmulr, clmulh

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Add comment about div/rem

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho commented:
Folks, we wrote a probe to double-check all latencies in the scheduler and updated the latencies accordingly. We tested it on two different boards to confirm the numbers.

The only outlier was idiv/irem, which was reported to have a latency of 3-4 in our experiments, so we went with the worst-case value shown in the C908 manual. I added a TODO to revisit this later. 

I also included the latencies for clmul, clmulr, and clmulh, which were missing from the first version of this PR.

asb commented on llvm/lib/Target/RISCV/RISCVSchedSpacemitX60.td:
```diff
@@ -0,0 +1,353 @@
+
+// Integer division/remainder
+// TODO: our experiments show that the latency of div is 4, which
+// seems too low. We used the worst case latency from the C908 instead.
```
I think this is too much detail for a bad test. Better to just say "TODO: Latency set based on C908 datasheet and hasn't been confirmed experimentally."

mikhailramalho committed:
Fix comment

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mshockwave commented:
> so we went with the worst-case value shown in the C908 manual

I think this is reasonable for integer divisions.

preames reviewed (commented):
Looks like this is converging with the feedback from @zqb-all (Thanks!).  Minor comment only.

preames reviewed (approved):
LGTM

zqb-all commented:
> Folks, we wrote a probe to double-check all latencies in the scheduler and updated the latencies accordingly. We tested it on two different boards to confirm the numbers.

Hi @mikhailramalho  , is "probe" referring to "llvm-exegesis", or does it refer to another independent program?

preames commented:
> > Folks, we wrote a probe to double-check all latencies in the scheduler and updated the latencies accordingly. We tested it on two different boards to confirm the numbers.
> 
> Hi @mikhailramalho , is "probe" referring to "llvm-exegesis", or does it refer to another independent program?

From context in offline discussion, this was an ad-hoc mix of llvm-exegesis where it seemed to produce sane results, and custom assembly snippets.  

There's definitely room for error here; this type of micro-architectural exploration is hard and error prone.  I approved this mostly based on the net perf results, not any expectation that every number for every instruction was correct.

I'd encourage you to make suggestions for improvements.  Ideally in the form of pull requests, but if you want to drop comments here, Mikhail or I can follow up.

mikhailramalho commented:
> > Folks, we wrote a probe to double-check all latencies in the scheduler and updated the latencies accordingly. We tested it on two different boards to confirm the numbers.
> 
> Hi @mikhailramalho , is "probe" referring to "llvm-exegesis", or does it refer to another independent program?

Hi @zqb-all, it's a custom probing tool that we plan to share soon
